README.md
# Instant Search Plugin for Zim Wiki
Search as you type in Zim, in similar manner to OneNote Ctrl+E.

When you hit Ctrl+E, small window opens, you can type in. As you type third letter, every page that matches your search is listed. You can walk through by <kbd>Up</kbd>/<kbd>Down</kbd> arrows or <kbd>Home</kbd>/<kbd>End</kbd>/<kbd>Page Up</kbd>/<kbd>Page Down</kbd> keys, hit <kbd>Enter</kbd> to stay on the page, or <kbd>Esc</kbd> to cancel.
Much quicker than current Zim search.

## Working with & Feedback
Known to work on:
 
* Ubuntu 15.10+ (still working on 20.04)
* Win 7 Zim 0.63+
* Debian 8.9 Zim 0.62+

I'd be glad to hear from you if it's working either here in the issues or in the original bug on [launchpad](https://bugs.launchpad.net/zim/+bug/1409626).

With old Zim 0.68 you may want to use the [last release](https://github.com/e3rd/zim-plugin-instantsearch/releases/tag/1.04) which is 0.68 compatible.
### Installation
Same as for the other plugins.
* Put the instantsearch.py into the plugins folder
  * something like %appdata%\zim\data\zim\plugins in Win, or /~/.local/share/zim/plugins/ in Linux
* You enable the plugin in Zim/Edit/Preferences/Plugins/ check mark Instant search.
* Type Ctrl+E and see if it's working, or report it here

## Demonstration on YouTube
Wanna see how it looks in action? In this example, I just search for the string "linux f" twice.
[![Demonstration](https://img.youtube.com/vi/nB2SfxDhEoM/0.jpg)](https://www.youtube.com/watch?v=nB2SfxDhEoM)

## Notes
* Prepend your string with an exclamation mark `!` to search in page names only
* Pages are found containing every piece of string you write, ex: `tour hou` will match page containing words `contour` and `silhouette`
* Favourizes page names, headers and exact query string matches, those are ordered first.
* More reliable than current version of the internal Zim search where the query `economical` is not recognized if the part of the text is bold: `economi**cal**` (however highlighting works great), if a link is inserted in the middle: `economi[[inserted link]]cal` or if the query is hidden in the link: `[[http://economical.example.com|link]]`.

# Copyright and License
Edvard Rejthar, [CSIRT.cz](https://csirt.cz), released under [LICENSE](LICENSE).


CHANGELOG
# 1.3.1 (unreleased)
* right position fix #50

# 1.3 (2022-12-15)
* part of the query might be divided between the page name and the page contents search #45
* search state is remembered (Ctrl+E will show last results)
* license info
* fix: open_when_unique parameter would not make the dialog re-close immediately any more
* fix: notebook located at the Windows drive root

# 1.2.1 (2021-11-02)
* main window resizing fix #46
* Zim 0.74 support #47

# 1.2 Python3.6 (2021-08-02)
* CHANGED: Check plugin preferences, some of them (using cache) resets to their defaults.
* preview: When navigating search menu, the page is previewed first after 150 ms by default and open after 1 500 ms. When page is not very small, navigation stuck when displaying a new page.
* results are being updated every 200 ms, not instantly, so that the search process might take about 30 % less time
* window title shows if the search has finished
* external fulltext search
    * internal Zim search dropped
    * wildcard option removed
    * searched texts are temporarily cached, drastically improving speed
* geometry fix, stays on position when results are longer than the dialog width
* page base name is in bold in results
* good caret behaviour
* good keyboard navigation
* fixed title search ('!t' would not match 'Other:Test' since 't' was incorrectly consumed in 'Other')

# 1.1 last Python3.5- (2020-06-09)

# 1.04 0.68 compatible (2018-06-18)
New config options, Tab and Shift+Tab for moving caret

# 1.03 Titles first (2017-01-16)
Since I made changes in this commit, the searching become even more smooth for me. The page titles are given the priority now to be ordered in the beginning.

# 1.02 History excluded (2016-08-27)
Traversing the search results should not fulfill your Zim history now ;)

# 1.01 #13 (2016-08-27)
many bug fixes, plugin is more reliable

# 1.0 Instant search for Zim (2016-06-03)
* For first few keystrokes, there is ultraquick search in page titles and then a reliable internal zim search is processed.
* more reliable search (less bugs)
* if I write "linux", the search for "linux" ends first, and than for "linu", "lin" in the background
* if I write "linu" and then "linux", the results are updated only, they are not cleared.
* if I write "linux" and then I paste from clipboard word "liposuction", the results are updated only, not cleared. So that if there is a page with both words "linux" and "liposuction", the page doesn't vanish for a while from menu. You wouldn't say you need this feature but I am convinced that if it disappeared, you would be confused and disappointed.
* if the searched string occures only in the page title and the only as one of the parent names, the results gets omitted. It's useful. When I search for "linux", I want to see page "os:linux", but not "os:linux:technologies:mydisk:foo:bar".
* While traversing the results by arrows, the page is displayed with tiny delay so that you can quickly traverse all the list.

tests.py
from collections import defaultdict
from unittest import TestCase, main

import gi

gi.require_version('Gtk', '3.0')

from instantsearch import SearchController, _MenuItem

cached_titles = [
    'Journal',
    'Journal:2021',
    'Journal:2021:12',
    'Journal:foo',
    'Journal:foo:bar',
    'Journal:foo:bar:fourth',
    'test',
    'Journal:test',
    'foo test',
    'foo (test)'
]


class TestSearch(TestCase):
    def _search(self, query, expected):
        menu = defaultdict(_MenuItem)
        SearchController.header_search(query, menu, cached_titles)
        self.assertListEqual(expected, [*menu])

    def test_header(self):
        self._search("foo", ['Journal:foo', 'Journal:foo:bar', 'Journal:foo:bar:fourth', 'foo test', 'foo (test)'])
        self._search("tes", ['test', 'Journal:test', 'foo test', 'foo (test)'])


if __name__ == '__main__':
    main()


typesearch.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#
# Search as you Type
#
#importy
import ConfigParser
from Tkinter import Tk, Entry, Label, StringVar
import re
import select
import subprocess
import sys
import termios
import time
import tty
import time


import gtk

from zim.plugins import PluginClass, extends, WindowExtension
from zim.actions import action
from zim.notebook import Path
from zim.gui.widgets import ui_environment, MessageDialog, Dialog

import logging

logger = logging.getLogger('zim.plugins.typesearch')
with open("/tmp/test.tmp","w+") as f:
    f.write(str(logger.manager.__dict__))


import threading
from functools import wraps


def delay(delay=0.):
    """
    Decorator delaying the execution of a function for a while. (http://fredericiana.com/2014/11/14/settimeout-python-delay/)
    """
    def wrap(f):
        @wraps(f)
        def delayed(*args, **kwargs):
            timer = threading.Timer(delay, f, args=args, kwargs=kwargs)
            timer.start()
        return delayed
    return wrap

class TypesearchPlugin(PluginClass):

    plugin_info = {
		'name': _('EDVARD'), # T: plugin name
		'description': _('''\
XXX

(V0.1)
'''), # T: plugin description
		'author': "Edvard Rejthar"
		#'help': 'Plugins:Due date',
	}


@extends('MainWindow')
class TypesearchMainWindowExtension(WindowExtension):

    uimanager_xml = '''
    <ui>
    <menubar name='menubar'>
            <menu action='tools_menu'>
                    <placeholder name='plugin_items'>
                            <menuitem action='typesearch'/>
                    </placeholder>
            </menu>
    </menubar>
    </ui>
    '''


    gui = "";


    @action(_('_Typesearch'), accelerator='<ctrl>e') # T: menu item
    def typesearch(self):
        #with open("/tmp/test.tmp","w+") as f:
        #    f.write(str(self.window.__dict__))
        #    f.write(str(self.window.ui.__dict__))
            
          #DAT GUI WIDTH a pozicovat doprava:
            #self.window.windowpos': (0, 24),
            #self.window.windowsize. (1920, 1056),
            

        #init
        self.query = "" #prikaz uzivatele
        self.caret = {'pos':0, 'altPos':0, 'text':""}  #pozice kurzoru
        self.matches = [] # XX lze sem dat recentne pouzite

        self.gui = Tk()
        Label(self.gui, text="Typesearch (if 1st letter is !, search in page titles only):").pack()
        self.gui.bind('<Up>', self.move)
        self.gui.bind('<Down>', self.move)
        self.gui.bind('<Enter>', self.move)
        self.gui.bind('<Return>', self.move)
        self.inputText = StringVar()
        self.inputText.trace("w", self.change)
        self.entry = Entry(self.gui, width=40, textvariable=self.inputText)
        self.entry.pack()
        self.entry.focus_set()
        self.labelText = StringVar()
        self.label = Label(self.gui, textvariable = self.labelText, justify = "left")
        self.label.pack()

        #self.gui.wm_attributes("-topmost", 1)

        self.gui.bind('<Up>', self.move) # XXX
        self.gui.bind('<Down>', self.move) # XXX
        self.gui.mainloop()

    
    lastQuery = ""
    lastPage = ""
    pageTitleOnly = False
    menu = []
    #queryTime = 0    

    def change(self,one, two, text):
        self.query = self.inputText.get() #self.entry.get() + event.char

        if len(self.query) < 3 or self.query == self.lastQuery:
            return

        self.lastQuery = self.query

        if self.query[:1] == "!": #prvni znak vykricnik - hleda se nazev stranky
            self.pageTitleOnly = True
            self.query = self.query[1:]
        else:
            self.pageTitleOnly = False


        #self.queryTime = int(round(time.time() * 1000))
        queryCheck = self.query
        self.gui.after(200, lambda: self.search(queryCheck))

    #@delay(0.2)
    def search(self,queryCheck):
        if self.query == "" or queryCheck != self.query: #mezitim jsme pripsali dalsi pismenko
            print("STORNO")
            return
        else:
            print("NON STORNO",self.query,queryCheck)
            #return
        #print millis
        #print("\x1B[3m" + (self.query if self.query else " ** character description **") + "\x1B[23m")# italikem vypsat prikaz
        self.menu = [] #mozne prikazy uzivatele
        self.caret['altPos'] = 0 #mozne umisteni karetu - na zacatek

        call = "zim --search Notes '*" + self.query + "*'"
        print(call)
        process = subprocess.Popen(call, stdout=subprocess.PIPE, shell=True)            
        self.matches = str(process.communicate()[0]).split("\n")[:-1] #, "utf-8"

        print("matches",self.matches)
        for option in self.matches:
            if self.pageTitleOnly and self.query not in option: # hledame jen v nazvu stranky
                continue

            self.menu.append(option) #zaradit mezi moznosti
            if option == self.caret['text']: #karet byl na tehle pozici, pokud se zuzil vyber, budeme vedet, kam karet opravne umistit
                self.caret['altPos'] = len(self.menu)-1
        self.displayMenu()

    def displayMenu(self):
        #osetrit vychyleni karetu
        if self.caret['pos'] < 0 or self.caret['pos'] > len(self.menu)-1: #umistit karet na zacatek ci konec seznamu
            self.caret['pos'] = self.caret['altPos']

        #vypsat self.menu
        #print("caret:" + str(caret['pos']))

        text = ""
        for i,item in enumerate(self.menu):
            if i == self.caret['pos']: #karet je na pozici
                self.caret['text'] = item
                text += '*' + item + "\n"#vypsat moznost tucne
            else:
                try:
                    text += item + "\n"
                except:
                    text += "CHYBA\n"
                    text += item[0:-1] + "\n"

        self.labelText.set(text)
        
        page = self.caret['text']
        if page and page != self.lastPage:
            self.lastPage = page
            print("page",page)
            #subprocess.Popen('zim Notes "'+page+'"', shell=True)
            self.window.ui.open_page(Path(page))
            
            # krade focus po pet vterin, abych mezitim mel nahledy otevrenych oken zimu;
            #  jestli z toho bude plugin, tak tahle kulisarna snad zmizi, protoze si bude se zimem povidat interne
            #for i in range(50,5000,50):
            #    self.gui.after(i, lambda: self.entry.focus_force())

        print("len", len(self.menu))
        print("len?", len(self.menu) == 1)
        if len(self.menu) == 1:
            if self.lastPage != page:
                self.window.ui.open_page(Path(page))
            self.gui.after(200, lambda: self.gui.destroy()) # kdyz se zaviralo hned, Python hazel allocation error
            #self.gui.quit()
            pass
            #subprocess.Popen('wmctrl -a zim', shell=True)

    def move(self,event):
        if event.keysym == "Up":
            self.caret['pos'] -= 1

        if event.keysym == "Down":
            self.caret['pos'] += 1

        if event.keysym == "Enter" or event.keysym == "Return":
            #self.menu = [self.menu[self.caret['pos']]] #launch command at caret
            self.gui.destroy()
            #exit(0)
            pass

        self.displayMenu()
        return

        # tadyto jsem vubec neudelal
        #res.configure(text = "levo!!")
        #vyhodnotit moznost
        if len(self.menu) != 1: #stale mame vic moznosti
            key = getkey()

            if ord(key) == 27: #escape sequence
                print("Hit escape for exit...") #pokud jsem predtim napsal sipku, dalsi 2 znaky automaticky cekaji
                key = getkey()
                if ord(key) == 27: #dvojity escape - ukonceni
                    exit()
                elif ord(key) == 91: #sipka
                    key = getkey()
                    if ord(key) == 65: #sipka nahoru
                        self.caret['pos'] -= 1
                    if ord(key) == 66: #sipka dolu
                        self.caret['pos'] += 1
                elif ord(key) == 79: #home/end
                    key = getkey()
                    if ord(key) == 72: #home
                        self.caret['pos'] = 0
                    if ord(key) == 70: #end
                        self.caret['pos'] = len(self.menu) -1
                else:
                    print("NIC!")
            #elif(key == '\x08' or key == '\x7f'):#backspace
            #    self.caret['pos'] = -1
            #    self.query = self.query[:-1]
            elif(ord(key) == 13): #enter
                self.menu = [self.menu[self.caret['pos']]] #launch command at caret
            #elif(key == '\x03' or key == '\x04'):#interrupt  (zel nefunguje)
            #    exit()
            elif(49 <= ord(key) <= 57):#cisla spousti prikaz na danem radku
                self.menu = [self.menu[ord(key)-49]]
                #query = self.menu[ord(key)-49] #launch command at number
            else:
                print("ZDE")
                self.caret['pos'] = -1
                self.query += key
        self.displayMenu()
#Typesearch()
